build/common.sh

```diff
@@ -317,6 +317,10 @@ function kube::build::has_docker() {
   which docker &> /dev/null
 }
 
+function kube::build::has_ip() {
+  which ip &> /dev/null
```

Comment:
This matches the style of the other `has_...` functions, but technically `command -v ip` is preferable. Even `type ip` is better than `which` given `which` has some weird historical behavior.

I think if we want to change this, it should be a separate PR that changes all of them probably though.

Reply:
I'd suggest doing something like `ip -V | grep -o 'iproute2'`.. this not only ensures there's a binary named "ip", but it's actually the "ip" tool from the iproute2 package.

Reply:
We don't do this elsewhere and I don't think we should do it here.

If someone wishes to have a roughly compatible `ip` tool with different `-V` output (such as [iproute2mac](https://github.com/brona/iproute2mac)), that's fine. Let them.

Second guessing users to that degree is the wrong solution to the problem.

The right solution is relying on users having a reasonable environment.

Reply:
We do do this elsewhere, for example, with sed: https://github.com/kubernetes/kubernetes/search?utf8=%E2%9C%93&q=gsed&type=

This is attempting to solve a build environment break, and it would be wrong to break other users environments just to fix yours. We should either default to ifconfig, or verify that the "ip" binary is the one we intend to call. Blindly defaulting to "ip" isn't the right choice in my opinion.

Reply:
In the case of sed, it's feature-detection for a common machine configuration.

> It would be wrong to break other users environments just to fix yours.

Linux machines that have a tool named `ip` available to noninteractive shells which isn't *the* `ip` tool are already broken.
We have no obligation to write overtly defensive and complex code to handle the limitless possible broken environments users can have.
Defailting to `ip` is the right choice because it's the new, non-deprecated, alternative to `ifconfig`. Defaulting to software which multiple distros are planning to delete is clearly the wrong choice.

I can agree with adding an `osx` check in `has_ip` to claim `osx` never has the tool.
Would such a check be sufficient in your opinion?

Reply:
My concern isn't specific to darwin.. that's just an example. The concern is that we want to ensure the utility we are calling is the one we intend. iproute2 is deprecating net-tools? Great. Then let's make sure the utility we are calling is from iproute2. Even iproute2mac (if installed) on darwin would [match the grep above](https://github.com/brona/iproute2mac/blob/bca8a46108fd7ad55de1a976ff3f7978b116eee9/src/ip.py#L535-L537) and work. But if the "ip" binary doesn't claim to be from iproute2, then it's not the right binary, we can't guarantee that it will respond to `ip address` in the right format, and we should probably fail back to ifconfig.

Or, as I've also suggested, another solution would be refactoring getting the IP address into a function of it's own, and then verifying that you have a valid IP address.

Reply:
Again, we can't handle every imaginary scenario. It's okay to declare some scenarios as broken dev environments we can ignore.

I think having a bad `ip` binary on linux falls well within that category.
Being cautious/defensive to that degree seems unreasonable to me. It's not based on any real-world linux distros I can think of.

Perhaps a second opinion from an owner of this code, such as @ixdy or @lavalamp would help resolve our apparent disagreement.

Reply:
> It's okay to declare some scenarios as broken dev environments we can ignore.

I'm glad to have other voices contribute to this, but the above statement is something I'd caution against. This would be the equivalent of suggesting that you should `alias ifconfig="ip address"` and be done with it. We want to provide high quality contributor experiences across the board :)

Reply:
@cblecker I just feel like we are getting into the weeds. We do not verify all the userland binaries the scripts shell out to. `which ip` would fail, by default, on a Darwin machine and defer to ifconfig, and the Linux distributions would work flawlessly with both ip and ifconfig installed. 

This style of functionality already exists in this file for the ensure_tar function.

https://github.com/rphillips/kubernetes/blob/4c00b6ed5eaabf938697741d3dda65b52146b9ff/build/common.sh#L295-L314

Reply:
Let's look at this another way then.. Is there a drawback to using `ip -V | grep -o 'iproute2'` instead of `which ip`? I hear that you both feel it's unnecessary, but is there any harms of going that way?

Reply:
Just to explain my position clearly @cblecker. Again, I'm not blocking because getting this change in at all is worth it, but I still think that it would be cleaner code to change the has_ip function to be `! kube::build::is_osx && command -v ip &>/dev/null`

The examples you've given of other commands where the version is checked are not relevant for the `ip` utility. sed and awk are checked because we *know* that there are machines out there without gnu compatible copies of them (such as busybox etc). The reason their version/features are being detected is because we *know* there's a possible problem.

Your argument here is not based on any real possibility. There are no linux machines we *know* of that include a faulty `ip` command. I can use `dnf provides *bin/ip` to tell me that no other fedora package includes the command. If you have evidence to the contrary, please present it, but until you do your argument is not based in reality.

> but is there any harms of going that way?

Writing more complicated code based on what-ifs that have 0 supporting evidence is how you end up with less readable, more complicated, and in general worse code.

The drawback is that this adds unneeded complexity. Someone who comes after will not be sure why this check was added without digging through discussion. It differs from the rest of the codebase (where e.g. [kubenet](https://github.com/kubernetes/kubernetes/blob/v1.8.0/pkg/kubelet/network/kubenet/kubenet_linux.go#L357) already relies on "ip" on linux being "ip").

The code complexity it adds is small, but I don't think we should do things based on "yeah, maybe it's totally pointless, but it's only this little thing". Each bit of pointless code adds up and leads to a worse codebase.

Reply:
This tweak makes sense to me... thoughts @cblecker?

Reply:
@euank If we are concerned about readability, I'd support adding a comment to describe that we are verifying that the ip command is the one provided by iproute2. The argument could be made that even exempting darwin would need a comment as to why we are doing it.

That said, in my opinion, validating that this is the utility we intend to call is not pointless. The point is to ensure that the utility we are calling not only exists, but serves the intended function as we are passing it's output directly into an environment variable.

As I mentioned [here](https://github.com/kubernetes/kubernetes/pull/53500#discussion_r143041596), I think best case is a slight refactor to have an independent function that retrieves and validates the IP address output, before passing it into the docker env variable.

Reply:
> even exempting darwin would need a comment as to why we are doing it.

Sure. A comment of "ip is a standard Linux utility" should clearly convey why we aren't bothering to check on other distros.

> That said, in my opinion, validating that this is the utility we intend to call is not pointless. The point is to ensure that the utility we are calling not only exists, but serves the intended function

It is a well named standard linux utility. We don't check that `/bin/sh` is posix shell before running shell scripts. We don't check that `docker` is the docker cli (even though there's more evidence it might not be; on debian `docker` can actually be the system tray, [docker](https://packages.debian.org/jessie/x11/docker) which provides a command by the same name).
I've already referenced evidence that this is a standard utility, that there aren't incompatible competing implementations.

On modern Linux machines, running the level of validations that venerable `./configure` scripts run is pointless.
I don't think anything you said properly responds to my comment nor provides evidence contrary to my point.